### ISSUE TITLE
MQTT Connection Drop

### ISSUE THREAD

**Issue by rothnic** [EinEinfach/CaSSAndRA#73]
It appears for me that after some time the server stops receiving mqtt messages, even though I see updates in mqtt explorer. Might need to handle trying to setup subscriptions every so often or something along those lines.

I started the server around 9:15AM, then it looks to stop receiving data around 9:45AM.
<img width="851" alt="image" src="https://github.com/EinEinfach/CaSSAndRA/assets/452052/23906f3f-c7e1-4b19-8917-6826f5a9a8e2">

You can see I'm still getting regular updates in mqtt explorer:

<img width="1018" alt="image" src="https://github.com/EinEinfach/CaSSAndRA/assets/452052/17f8fb5b-f400-44f5-a492-4f35b7817317">

Haven't had a chance to dig into it, but one initial option might be to add a callback function for `on_disconnect` so we can at least see if that is happening.

**Comment by EinEinfach** [comment]
Very strange... sure that MQTT Explorer still get MQTT messages? I can't see any time stamps. If MQTT Explorer also lose connection you can try switch to other the MQTT library in ESP32 fw. Here is someone who had also problems with ESP32 MQTT
[https://forum.ardumower.de/threads/esp32-mqtt-funktioniert-nicht-24-7-solved.25274/post-62256](url) 

> Haven't had a chance to dig into it, but one initial option might be to add a callback function for on_disconnect so we can at least see if that is happening.

You get already a feedback in the app when connection get lost (State string chages to offline). At 9:47:28 state string changed his state to offline.

**Comment by rothnic** [comment]
Yeah mqtt explorer shows updates the whole time, constantly updating. Shows 6:35PM in the app. The odd thing is when running a task, I haven't seen any drop outs.

What I mean by adding the on_disconnect is you could log it, but more importantly we could also handle reconnecting. I can see where we decided it was offline, but I can't tell that it is reconnecting. If I restart the server it immediately starts showing the same frequency of updates I see in the mqtt explorer without any signs of intermittent or spotty data. So, it seems the client has to get into a bad state of some kind. 

Kind of brute force, but restarting the client after some time without updates would be a potential work around to try with some kind of back off.

**Comment by ShadedSelf** [comment]
Same here, I had to switch to HTTP.
I feel like this didnt happen before though, I cant remember when it started happening, it was a while ago, though maybe its always been like this.

**Comment by EinEinfach** [comment]
Ok, I will take a look at MQTT and implement on_disconnect callback

**Comment by rothnic** [comment]
> Ok, I will take a look at MQTT and implement on_disconnect callback

I'll do some investigating as well. Going to start with this just to see if on_disconnect actually gets triggered when I see the data stop coming through. If it is actually getting called, then we could maybe use that connection flag to detect when it is no longer connected, and re-initialize the mqtt client and subscriptions as an initial work around. 

```
def connect_mqtt(connect_data: dict()) -> mqtt:
    client_id = connect_data['MQTT'][0]['CLIENT_ID'] + str(random.randint(1, 1000))
    username = connect_data['MQTT'][1]['USERNAME'] 
    password = connect_data['MQTT'][2]['PASSWORD']
    mqtt_server = connect_data['MQTT'][3]['MQTT_SERVER']
    port = connect_data['MQTT'][4]['PORT']
    def on_connect(client, userdata, flags, rc):
        if rc == 0:
            logger.info('Backend: Subscriptions to the rover MQTT topics succsessfull')
        else:
            client.connection_flag = False
            logger.warning('Backend: MQTT connection to the rover lost. Code: '+str(rc))

    def on_disconnect(client, userdata, rc):
        logger.info(f"Backend: MQTT disconnecting reason {rc}")
        logger.debug(f"Backend: MQTT disconnect userdata {userdata}")
        client.connection_flag = False

    # Set Connecting Client ID
    client = mqtt.Client(client_id)
    client.connection_flag = False
    client.username_pw_set(username, password)
    client.on_connect = on_connect
    client.on_disconnect = on_disconnect
    try:
        client.connect(mqtt_server, port)
        client.connection_flag = True
        logger.info('Backend: Connection to the MQTT-Server succsessfull')
    except:
        logger.warning('Backend: Connection to the MQTT-Server failed')

    return client
```

**Comment by EinEinfach** [comment]
Good idea, good to know if on_disconnect will be triggered. Don't take to much time for refactoring code. I'm currently writing a new dataclass for MQTT connection.

**Comment by rothnic** [comment]
> Good idea, good to know if on_disconnect will be triggered. Don't take to much time for refactoring code. I'm currently writing a new dataclass for MQTT connection.

Sounds good, yeah hoping to just catch the disconnect at this point. I'll let you know if I see anything. In the past it seemed to stop showing messages after 30 min to an hour, so hopefully will know something here soon.

**Comment by rothnic** [comment]
Ok @EinEinfach , got the output from the on_disconnect method. 

It does get called, but surprisingly I see the same status code (rc=7) as I saw before making the client id more unique. I'm now second guessing what the disconnect codes mean and can't seem to find where that is clearly documented. I see documentation for the connection codes, but it seems these are different than the disconnect codes.

```
2023-09-20 11:45:44; DEBUG Backend: Calc data from state data frame
2023-09-20 11:45:50; INFO Backend: RX topic:ardumower/Ardumower/state message:b'{"battery_voltage":28.56,"position":{"x":11,"y":22.14,"delta":3.14,"solution":2,"age":0.4,"accuracy":0.02,"visible_satellites":28,"visible_satellites_dgps":23,"mow_point_index":0},"target":{"x":13.52,"y":22.13},"job":2,"sensor":0,"amps":-0.01,"map_crc":2111653}'
2023-09-20 11:45:50; DEBUG Backend: Calc data from state data frame
2023-09-20 11:45:56; INFO Backend: RX topic:ardumower/Ardumower/state message:b'{"battery_voltage":28.55,"position":{"x":11,"y":22.14,"delta":3.14,"solution":2,"age":0.35,"accuracy":0.02,"visible_satellites":32,"visible_satellites_dgps":25,"mow_point_index":0},"target":{"x":13.52,"y":22.13},"job":2,"sensor":0,"amps":-0.01,"map_crc":2111653}'
2023-09-20 11:45:56; DEBUG Backend: Calc data from state data frame
2023-09-20 11:46:15; INFO Backend: MQTT disconnecting reason 7
2023-09-20 11:46:15; DEBUG Backend: MQTT disconnect userdata None
2023-09-20 11:46:16; INFO Backend: Subscriptions to the rover MQTT topics succsessfull
2023-09-20 11:46:26; WARNING Backend: Could not connect to the rover. State set to offline
2023-09-20 11:51:15; INFO Backend: Writing State-Data to the file
2023-09-20 11:51:15; INFO Backend: State data are saved successfully in state.pickle
2023-09-20 11:51:16; INFO Backend: Calced state data are saved successfully in calcstate.pickle
```

I receive a MQTT message at `2023-09-20 11:45:56`, then we seem to trigger the on_connect method at `2023-09-20 11:46:16;`, which might be a little bit misleading of a log message that we have just setup subscriptions. Then at `2023-09-20 11:46:26;` it is considered offline.

I then kill the server, and I'm right back to receiving mqtt messages.

**Comment by rothnic** [comment]
I'm running a mosquitto mqtt server, which has primarily been used for home assistant. I checked the logs of the mosquitto server and found some logs and pulled key parts out. It shows a socket disconnect error, which who knows why this happens. 

We seem to reconnect one time, but then another time it doesn't. One thing I saw mentioned in some related errors that were discussed was the ping/keep alive on the client can sometimes be the cause of this, but didn't see a clear pattern for why. Might be where we need to look at the timing, but ultimately we should try to avoid the state where we disconnect and don't handle it within cassandra. Just knowing that we can restart the server and reconnect tells me we should be able to setup the connection again with possibly some kind of back off period.

Here is where cassandra connects and subscribes:

```
1695226856: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695226862: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695226868: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695226872: New client connected from 10.0.0.15 as Ardumower850 (c1, k60, u'mqtt').
1695226872: Sending CONNACK to Ardumower850 (0, 0)
1695226872: Received SUBSCRIBE from Ardumower850
1695226872:     ardumower/Ardumower/state (QoS 0)
1695226872: Ardumower850 0 ardumower/Ardumower/state
1695226872: Sending SUBACK to Ardumower850
1695226872: Received SUBSCRIBE from Ardumower850
1695226872:     ardumower/Ardumower/stats (QoS 0)
1695226872: Ardumower850 0 ardumower/Ardumower/stats
1695226872: Sending SUBACK to Ardumower850
1695226872: Received SUBSCRIBE from Ardumower850
1695226872:     ardumower/Ardumower/props (QoS 0)
1695226872: Ardumower850 0 ardumower/Ardumower/props
1695226872: Sending SUBACK to Ardumower850
1695226872: Received SUBSCRIBE from Ardumower850
1695226872:     ardumower/Ardumower/online (QoS 0)
1695226872: Ardumower850 0 ardumower/Ardumower/online
1695226872: Sending SUBACK to Ardumower850
1695226872: Received PUBLISH from Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/command', ... (10 bytes))
1695226872: Sending PUBLISH to Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/command', ... (10 bytes))
1695226872: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
```

couple ping request and response examples:

```
1695226932: Received PINGREQ from Ardumower850
1695226932: Sending PINGRESP to Ardumower850
1695226992: Received PINGREQ from Ardumower850
1695226992: Sending PINGRESP to Ardumower850
```

Disconnect and reconnect
```
1695228314: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228314: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228314: Received PINGREQ from Ardumower850
1695228314: Sending PINGRESP to Ardumower850
1695228320: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228320: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228326: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228326: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228332: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228332: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228338: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228338: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228341: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
1695228341: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
1695228344: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228344: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228350: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228350: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228356: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228356: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228362: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228362: Sending PUBLISH to Ardumower850 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228362: Socket error on client Ardumower850, disconnecting.
1695228368: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228374: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228376: New client connected from 10.0.0.15 as Ardumower850 (c1, k60, u'mqtt').
1695228376: Sending CONNACK to Ardumower850 (0, 0)
1695228380: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228386: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228392: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228398: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228402: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
1695228404: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228410: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228416: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228422: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228428: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695228434: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228436: Received PINGREQ from Ardumower850
1695228436: Sending PINGRESP to Ardumower850
1695228440: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228446: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695228452: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
```


Final disconnect
```
1695229637: Received PINGREQ from Ardumower850
1695229637: Sending PINGRESP to Ardumower850
1695229640: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229646: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229652: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695229658: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229664: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229670: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229676: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229682: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229687: Socket error on client Ardumower850, disconnecting.
1695229687: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
1695229688: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695229694: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
```

**Comment by rothnic** [comment]
These logs also explain why i was getting the disconnects before. Didn't realize i had set the mqtt client id to Ardumower, rather than Cassandra. No wonder I had conflicts, but having unique client ids appears to be best practice anyways.

**Comment by rothnic** [comment]
> I have no time in the next days. If you want you can finish my idea.

Sure, I can take a look at it.

As for an update, I happened to lose my network connection for a decent period of time just a bit ago on the device running cassandra at the moment. It seemed to detect it as offline, then it connected back and started working with no problem. So, I think the proposed changes to setting up the subscriptions fixes the problem, but could use some more testing.

```
2023-09-20 15:19:41; DEBUG Backend: Calc data from state data frame
2023-09-20 15:19:42; INFO Backend: RX topic:ardumower/Ardumower/state message:b'{"battery_voltage":28.58,"position":{"x":11,"y":22.14,"delta":3.14,"solution":0,"age":0.54,"accuracy":0.11,"visible_satellites":42,"visible_satellites_dgps":32,"mow_point_index":0},"target":{"x":13.52,"y":22.13},"job":2,"sensor":0,"amps":-0.01,"map_crc":2111653}'
2023-09-20 15:19:42; DEBUG Backend: Calc data from state data frame
2023-09-20 15:20:13; WARNING Backend: Could not connect to the rover. State set to offline
2023-09-20 15:20:31; INFO Backend: Writing State-Data to the file
2023-09-20 15:20:32; INFO Backend: State data are saved successfully in state.pickle
2023-09-20 15:20:32; INFO Backend: Calced state data are saved successfully in calcstate.pickle
2023-09-20 15:20:32; INFO Backend: Writing Statistics-Data to the file
2023-09-20 15:20:33; INFO Backend: Statistics data are saved successfully in stats.pickle
2023-09-20 15:20:33; INFO Backend: Calced statistics data are saved successfully in calcstats.pickle
2023-09-20 15:20:38; INFO Backend: MQTT disconnecting reason 7
2023-09-20 15:20:38; DEBUG Backend: MQTT disconnect userdata None
2023-09-20 15:30:33; INFO Backend: Writing State-Data to the file
2023-09-20 15:30:33; INFO Backend: State data are saved successfully in state.pickle
2023-09-20 15:30:34; INFO Backend: Calced state data are saved successfully in calcstate.pickle
2023-09-20 15:30:34; INFO Backend: Writing Statistics-Data to the file
2023-09-20 15:30:34; INFO Backend: Statistics data are saved successfully in stats.pickle
2023-09-20 15:30:34; INFO Backend: Calced statistics data are saved successfully in calcstats.pickle
2023-09-20 15:40:34; INFO Backend: Writing State-Data to the file
2023-09-20 15:40:35; INFO Backend: State data are saved successfully in state.pickle
2023-09-20 15:40:35; INFO Backend: Calced state data are saved successfully in calcstate.pickle
2023-09-20 15:40:35; INFO Backend: Writing Statistics-Data to the file
2023-09-20 15:40:35; INFO Backend: Statistics data are saved successfully in stats.pickle
2023-09-20 15:40:35; INFO Backend: Calced statistics data are saved successfully in calcstats.pickle
2023-09-20 15:41:43; INFO Backend: Client connection to mqtt server successful
2023-09-20 15:41:43; INFO Backend: Setting up subscriptions and message handling.
2023-09-20 15:41:43; INFO Backend: Subscribed to ardumower/Ardumower/state
2023-09-20 15:41:43; INFO Backend: Subscribed to ardumower/Ardumower/stats
2023-09-20 15:41:43; INFO Backend: Subscribed to ardumower/Ardumower/props
2023-09-20 15:41:43; INFO Backend: Subscribed to ardumower/Ardumower/online
2023-09-20 15:41:45; INFO Backend: RX topic:ardumower/Ardumower/state message:b'{"battery_voltage":28.58,"position":{"x":11,"y":22.14,"delta":3.14,"solution":2,"age":0.52,"accuracy":0.02,"visible_satellites":36,"visible_satellites_dgps":31,"mow_point_index":0},"target":{"x":13.52,"y":22.13},"job":2,"sensor":0,"amps":-0.02,"map_crc":2111653}'
2023-09-20 15:41:45; DEBUG Backend: Calc data from state data frame
2023-09-20 15:41:51; INFO Backend: RX topic:ardumower/Ardumower/state message:b'{"battery_voltage":28.59,"position":{"x":11,"y":22.14,"delta":3.14,"solution":2,"age":0.49,"accuracy":0.02,"visible_satellites":35,"visible_satellites_dgps":30,"mow_point_index":0},"target":{"x":13.52,"y":22.13},"job":2,"sensor":0,"amps":-0.01,"map_crc":2111653}'
```

There here are the mqtt logs from the disconnect:

```
1695241227: Sending PUBLISH to Ardumower722 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
1695241233: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695241233: Sending PUBLISH to Ardumower722 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695241238: Socket error on client Ardumower722, disconnecting.
1695241239: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695241242: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/stats', ... (607 bytes))
1695241245: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (261 bytes))
```

Then from where the cassandra client reconnected after some time and automatically resubscribed, and started updating the state again.

```
1695242493: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695242499: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695242503: New client connected from 10.0.0.15 as Ardumower722 (c1, k60, u'mqtt').
1695242503: Sending CONNACK to Ardumower722 (0, 0)
1695242503: Received SUBSCRIBE from Ardumower722
1695242503:     ardumower/Ardumower/state (QoS 0)
1695242503: Ardumower722 0 ardumower/Ardumower/state
1695242503: Sending SUBACK to Ardumower722
1695242503: Received SUBSCRIBE from Ardumower722
1695242503:     ardumower/Ardumower/stats (QoS 0)
1695242503: Ardumower722 0 ardumower/Ardumower/stats
1695242503: Sending SUBACK to Ardumower722
1695242503: Received SUBSCRIBE from Ardumower722
1695242503:     ardumower/Ardumower/props (QoS 0)
1695242503: Ardumower722 0 ardumower/Ardumower/props
1695242503: Sending SUBACK to Ardumower722
1695242503: Received SUBSCRIBE from Ardumower722
1695242503:     ardumower/Ardumower/online (QoS 0)
1695242503: Ardumower722 0 ardumower/Ardumower/online
1695242503: Sending SUBACK to Ardumower722
1695242505: Received PUBLISH from Ardumower (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
1695242505: Sending PUBLISH to Ardumower722 (d0, q0, r0, m0, 'ardumower/Ardumower/state', ... (262 bytes))
```

**Comment by EinEinfach** [comment]
Good. How many changes did you make in code? If this just some rows make a PR and we have a solution for now. Then I can take more time for recreating of connection classes

**Comment by rothnic** [comment]
> Good. How many changes did you make in code? If this just some rows make a PR and we have a solution for now. Then I can take more time for recreating of connection classes

Yeah, that is what I was thinking. This is a pretty small rework, which the new class would need to implement as well. It might be easier to refactor the new class starting from a working version and I can get that pull request in for the fix pretty quickly.

This morning it still is connected and working, so yeah think this approach seems to work.

**Comment by rothnic** [comment]
Closed due to merging of #74, but if anyways sees more mqtt issues in the meantime, let me know.
